### PR TITLE
chore: extract write logic into service layer

### PR DESCRIPTION
**Pull request by sfirke**
PR 4 of a series related to https://github.com/sfirke/meutch/issues/341.

### CODE REVIEW

File: app/services/circle_service.py [hunk 1]
```diff
@@ -117,6 +219,173 @@ def cancel_join_request(circle_id, user_id):
...
+    if not circle.is_admin(acting_user):
+        raise AuthorizationError("You must be an admin to remove members.")
+
+    if member_user not in circle.members:
+        raise ConflictError("User is not a member of this circle.")
+
+    if member_user == acting_user:
+        raise InvalidActionError("Use the leave circle button to remove yourself.")
+
+    if circle.is_admin(member_user) and _count_circle_admins(circle.id) <= 1:
```
Comment (sfirke) [inline app/services/circle_service.py]: I don't think this condition can be reached

File: app/services/circle_service.py [hunk 2]
```diff
@@ -135,6 +404,8 @@ def toggle_admin(circle, user_id, acting_user, action):
     if action == "add":
         is_admin = True
     elif action == "remove":
+        if user_member.is_admin and _count_circle_admins(circle.id) <= 1:
```
Comment (sfirke) [inline app/services/circle_service.py]: Same problem, condition can't be reached because user_member and acting_user are both admins

File: app/models.py [hunk 3]
```diff
@@ -170,15 +170,23 @@ def get_active_loans_as_borrower(self):
         """Returns items user is currently borrowing"""
         return (
             Item.query.join(LoanRequest)
-            .filter(LoanRequest.borrower_id == self.id, LoanRequest.status == "approved")
+            .filter(
+                Item.is_giveaway.is_(False),
```
Comment (sfirke) [inline app/models.py]: Is this covering a legacy possible condition where a loanrequest could be a giveaway? Don't think we need to worry about that

File: app/models.py [hunk 4]
```diff
@@ -170,15 +170,23 @@ def get_active_loans_as_borrower(self):
...
             .all()
         )
 
     def get_active_loans_as_owner(self):
         """Returns items user is currently lending"""
         return (
             Item.query.join(LoanRequest)
-            .filter(Item.owner_id == self.id, LoanRequest.status == "approved")
+            .filter(
+                Item.is_giveaway.is_(False),
```
Comment (sfirke) [inline app/models.py]: Same as prior comment: Is this covering a legacy possible condition where a loanrequest could be a giveaway? Don't think we need to worry about that

File: tests/integration/test_routes.py [hunk 5]
```diff
@@ -767,6 +767,43 @@ def test_delete_modal_for_active_loan_shows_resolution_guidance(self, client, ap
             assert b"Mark the item returned or cancel the loan before deleting it." in response.data
             assert b"View Active Loan" in response.data
 
+    def test_complete_giveaway_backed_loan_is_blocked(self, client, app, auth_user):
```
Comment (sfirke) [inline tests/integration/test_routes.py]: Same, don't need to cover this possibility

File: tests/unit/services/test_loan_service.py [hunk 6]
```diff
@@ -180,6 +180,46 @@ def test_complete_loan_raises_auth_error_for_non_owner(self, app):
             with pytest.raises(AuthorizationError):
                 loan_service.complete_loan(loan, other.id)
 
+    def test_process_loan_decision_rejects_giveaway_backed_loan(self, app):
```
Comment (sfirke) [inline tests/unit/services/test_loan_service.py]: Remove? Legacy conflict condition

File: tests/unit/services/test_loan_service.py [hunk 7]
```diff
@@ -180,6 +180,46 @@ def test_complete_loan_raises_auth_error_for_non_owner(self, app):
...
+        with app.app_context():
+            owner = UserFactory()
+            borrower = UserFactory()
+            item = ItemFactory(owner=owner, is_giveaway=True, available=False)
+            loan = LoanRequestFactory(item=item, borrower=borrower, status="pending")
+
+            with pytest.raises(ConflictError, match="giveaway"):
+                loan_service.process_loan_decision(loan, owner.id, "approve")
+
+    def test_cancel_loan_request_rejects_giveaway_backed_loan(self, app):
```
Comment (sfirke) [inline tests/unit/services/test_loan_service.py]: Same, is that "giveaway backed loan"

File: tests/unit/services/test_loan_service.py [hunk 8]
```diff
@@ -180,6 +180,46 @@ def test_complete_loan_raises_auth_error_for_non_owner(self, app):
...
+        with app.app_context():
+            owner = UserFactory()
+            borrower = UserFactory()
+            item = ItemFactory(owner=owner, is_giveaway=True, available=False)
+            loan = LoanRequestFactory(item=item, borrower=borrower, status="pending")
+
+            with pytest.raises(ConflictError, match="giveaway"):
+                loan_service.cancel_loan_request(loan, borrower.id)
+
+    def test_complete_loan_rejects_giveaway_backed_loan(self, app):
```
Comment (sfirke) [inline tests/unit/services/test_loan_service.py]: Same

File: tests/unit/services/test_loan_service.py [hunk 9]
```diff
@@ -180,6 +180,46 @@ def test_complete_loan_raises_auth_error_for_non_owner(self, app):
...
+        with app.app_context():
+            owner = UserFactory()
+            borrower = UserFactory()
+            item = ItemFactory(owner=owner, is_giveaway=True, available=False)
+            loan = LoanRequestFactory(item=item, borrower=borrower, status="approved")
+
+            with pytest.raises(ConflictError, match="giveaway"):
+                loan_service.complete_loan(loan, owner.id)
+
+    def test_owner_cancel_rejects_giveaway_backed_loan(self, app):
```
Comment (sfirke) [inline tests/unit/services/test_loan_service.py]: Same

File: tests/unit/services/test_loan_service.py [hunk 10]
```diff
@@ -196,3 +236,19 @@ def test_extend_loan_raises_auth_error_for_non_owner(self, app):
 
             with pytest.raises(AuthorizationError):
                 loan_service.extend_loan(loan, other.id, date.today() + timedelta(days=7), "")
+
+    def test_extend_loan_rejects_giveaway_backed_loan(self, app):
```
Comment (sfirke) [inline tests/unit/services/test_loan_service.py]: Same

File: tests/unit/test_models.py [hunk 11]
```diff
@@ -226,6 +226,34 @@ def test_user_shared_circles_with_returns_empty_for_missing_user(self, app):
             assert viewer.shared_circles_with(None) == []
             assert viewer.shares_circle_with(None) is False
 
+    def test_get_active_loans_as_borrower_excludes_giveaway_items(self, app):
```
Comment (sfirke) [inline tests/unit/test_models.py]: LoanRequestFactory on the giveaway item? Another one of these edge cases I don't want to cover

File: tests/unit/test_models.py [hunk 12]
```diff
@@ -226,6 +226,34 @@ def test_user_shared_circles_with_returns_empty_for_missing_user(self, app):
...
+            giveaway_item = ItemFactory(owner=owner, is_giveaway=True, available=False)
+
+            LoanRequestFactory(item=loan_item, borrower=borrower, status="approved")
+            LoanRequestFactory(item=giveaway_item, borrower=borrower, status="approved")
+
+            borrowing = borrower.get_active_loans_as_borrower()
+
+            assert [item.id for item in borrowing] == [loan_item.id]
+
+    def test_get_active_loans_as_owner_excludes_giveaway_items(self, app):
```
Comment (sfirke) [inline tests/unit/test_models.py]: And this

**Review by sfirke — COMMENTED** [review]
Just the same edge case over and over that it's not worth covering
